components/tidb_query/src/rpn_expr/impl_string.rs

```diff
@@ -197,6 +198,24 @@ pub fn right(lhs: &Option<Bytes>, rhs: &Option<Int>) -> Result<Option<Bytes>> {
     }
 }
 
+#[rpn_fn(capture = [extra])]
+#[inline]
+pub fn upper(extra: &RpnFnCallExtra, arg: &Option<Bytes>) -> Result<Option<Bytes>> {
+    match arg.as_ref() {
+        Some(bytes) => {
+            if extra.ret_field_type.as_accessor().is_binary_string_like() {
```

Comment:
We should not check the return field type of Upper expression, but should check the child return filed type. We cannot do this in the current implementation and we will support it in recent.

Reply:
@lonng This can be resolved if we separate binary and utf8 methods into two. But how do think the idea of the separation?

Reply:
@breeswish Yes, we can resolve it by separate it into two signatures in the current implementation, but I think it’s away from our goal.
1. It’s better to keep consistent with TiDB in signature compatibility for maintaining purpose.
2. There is risk to make out signatures expand.

Reply:
@lonng

Reply:
@lonng TiDB itself need to be changed if signature is going to be changed :) I just want to make it consistent with other string functions because they all have binary and non-binary versions (and separating signatures is good for better vectorizations because the branch is moved to the outer of the loop).

Reply:
@breeswish agree with you that split it into two signatures.

Reply:
@lonng Looks like we can add an easy task, to separate signatures for int / uint as well! (we have these mappers at TiKV side, but they should be moved to TiDB side and assign different signatures)?

Reply:
@breeswish  
Do separation like this?
https://github.com/pingcap/tidb/blob/2b93239c66e008bfbd85e61d9e2756d5d40acc53/expression/builtin_string.go#L1046

Reply:
Yes, exactly! You can do it in future PRs. It's ok.

Reply:
I found the PR https://github.com/pingcap/tipb/pull/163. Does it mean I should follow the new name patterns?

Reply:
@wangwangwar It's up to you :) You can finish this PR first (since it is *correct*), and then open new PRs (in tipb, tidb and tikv) to improve `upper` to a better implementation.